include/proj/io.hpp

```diff
@@ -803,145 +886,143 @@ class PROJ_GCC_DLL DatabaseContext {
                                   bool &directDownload, bool &openLicense,
-                                  bool &gridAvailable) const;
+                                  bool &gridAvailable) const override;
 
     PROJ_INTERNAL std::string
     getAliasFromOfficialName(const std::string &officialName,
                              const std::string &tableName,
-                             const std::string &source) const;
+                             const std::string &source) const override;
 
     PROJ_INTERNAL bool isKnownName(const std::string &name,
-                                   const std::string &tableName) const;
+                                   const std::string &tableName) const override;
 
     PROJ_INTERNAL std::string getTextDefinition(const std::string &tableName,
                                                 const std::string &authName,
-                                                const std::string &code) const;
+                                                const std::string &code) const override;
 
     PROJ_INTERNAL std::vector<std::string>
     getAllowedAuthorities(const std::string &sourceAuthName,
-                          const std::string &targetAuthName) const;
+                          const std::string &targetAuthName) const override;
 
     PROJ_INTERNAL std::list<std::pair<std::string, std::string>>
     getNonDeprecated(const std::string &tableName, const std::string &authName,
-                     const std::string &code) const;
+                     const std::string &code) const override;
 
     //! @endcond
 
   protected:
-    PROJ_INTERNAL DatabaseContext();
+    PROJ_INTERNAL SQLiteDatabaseContext();
     INLINED_MAKE_SHARED
-    PROJ_FRIEND(AuthorityFactory);
+    PROJ_FRIEND(SQLiteAuthorityFactory);
 
   private:
     PROJ_OPAQUE_PRIVATE_DATA
-    DatabaseContext(const DatabaseContext &) = delete;
-    DatabaseContext &operator=(const DatabaseContext &other) = delete;
+    SQLiteDatabaseContext(const SQLiteDatabaseContext &) = delete;
+    SQLiteDatabaseContext &operator=(const SQLiteDatabaseContext &other) = delete;
 };
 
 // ---------------------------------------------------------------------------
 
-class AuthorityFactory;
-/** Shared pointer of AuthorityFactory. */
-using AuthorityFactoryPtr = std::shared_ptr<AuthorityFactory>;
-/** Non-null shared pointer of AuthorityFactory. */
-using AuthorityFactoryNNPtr = util::nn<AuthorityFactoryPtr>;
-
-/** \brief Builds object from an authority database.
+/** \brief Interface for building objects from an authority database.
  *
  * A AuthorityFactory should be used only by one thread at a time.
  *
  * \remark Implements [AuthorityFactory]
  * (http://www.geoapi.org/3.0/javadoc/org/opengis/referencing/AuthorityFactory.html)
  * from \ref GeoAPI
  */
-class PROJ_GCC_DLL AuthorityFactory {
+class PROJ_GCC_DLL AuthorityFactory  :
+        public util::nn_enable_shared_from_this<AuthorityFactory> {
```

Comment:
why this derivation for util::nn_enable_shared_from_this<AuthorityFactory> ? I had avoided deriving from std::enable_shared_from_this<> to avoid potential issues with MSVC and exporting interfaces of std:: namespace which used to cause issues in the past

Reply:
Same remark as a previous one, but with an increased criticality. You put in the interface really high level functionnalities. Implementing correctly some of those methods is really involved and requires good knowledge of the database model. It might - and will - change in a near future as well, so maintaining that for external implementations of this class will be a lot of work